stablehlo/reference/Ops.cpp

```diff
@@ -36,6 +36,18 @@ Tensor eval(ConstantOp op, ElementsAttr value) {
   return Tensor(value.cast<DenseElementsAttr>());
 }
 
+Tensor eval(ReshapeOp op, const Tensor &operand) {
+  // The fact that the interpreter stores the tensor internally as contiguous
+  // byte arrays allows to reshape the dimensions of tensor merely by modifying
+  // it's strides, while storing the same internal memory as that of the
+  // operand.
```

Comment:
Similarly to the spec proposal above, let's define op implementations from first principles, i.e. using only the definitions in the spec and APIs from `Tensor.h` and `Element.h` which implement these definitions. 

From that perspective, relying on exact representation of a tensor, i.e. something that the spec treats as implementation-defined (currently, this is implicit, #193 makes this explicit), doesn't sound right.

I think improving on this is fairly straightforward if we use the iterator that you proposed in #167 (with the modifications that I proposed in the comments), for example:

```
Tensor eval(ReshapeOp op, const Tensor &operand) {
  Tensor result(op.getType());
  for (auto operandIt = operand.index_begin(), resultIt = result.index_begin();
       operandIt != operand.index_end();
       ++operandIt, ++resultIt) {
    result.set(*it, operand.get(*it));
  }
  return result;
}
```

Reply:
Agreed. I will modify as suggested once https://github.com/openxla/stablehlo/pull/167 will land.